scalafix-core/shared/src/main/scala/scalafix/internal/patch/EscapeHatch.scala

```diff
@@ -118,261 +79,409 @@ class EscapeHatch(
+    def disableEscapes: Iterable[EscapeFilter] = escapeTree.values.flatten
   }
 
-  private sealed abstract class Toogle
-  private object Toogle {
-    case object Disable extends Toogle
-    case object Enable extends Toogle
-  }
+  private object AnnotatedEscapes {
+    private val SuppressWarnings = "SuppressWarnings"
+    private val SuppressAll = "all"
+    val OptionalRulePrefix = "scalafix:"
+
+    def apply(tree: Tree): AnnotatedEscapes = {
+      val builder = TreeMap.newBuilder[EscapeOffset, List[EscapeFilter]]
+
+      def addAnnotatedEscape(t: Tree, mods: List[Mod]): Unit = {
+        val start = EscapeOffset(t.pos.start)
+        val end = EscapeOffset(t.pos.end)
+        val rules = extractRules(mods)
+        val (matchAll, matchOne) = rules.partition(_._1 == SuppressAll)
+        val filters = ListBuffer.empty[EscapeFilter]
+
+        // 'all' should come before individual rules so that we can warn unused rules later
+        for ((_, rulePos) <- matchAll) {
+          val matcher = FilterMatcher.matchEverything
+          filters += EscapeFilter(matcher, rulePos, start, Some(end))
+        }
+        for ((rule, rulePos) <- matchOne) {
+          val unprefixedRuleName = rule.stripPrefix(OptionalRulePrefix)
+          val matcher = FilterMatcher(unprefixedRuleName)
+          filters += EscapeFilter(matcher, rulePos, start, Some(end))
+        }
 
-  private val unusedScalafixSupression = RuleName("UnusedScalafixSupression")
+        builder += (start -> filters.result())
+      }
 
-  private val unusedEnableWarning =
-    LintCategory
-      .warning(
-        "Enable",
-        "This comment would enable a rule that was not disabled (eg: typo in the rules)"
-      )
-      .withOwner(unusedScalafixSupression)
-
-  private val FilterDisable = "\\s?scalafix:off\\s?(.*)".r
-  private val FilterEnable = "\\s?scalafix:on\\s?(.*)".r
-  private val FilterExpression = "\\s?scalafix:ok\\s?(.*)".r
-
-  private def splitRules(rules: String): Set[String] =
-    rules.split("\\s+").toSet
-
-  def apply(tree: Tree, associatedComments: AssociatedComments): EscapeHatch = {
-    val escapes = List.newBuilder[Escape]
-    val unusedEnable = List.newBuilder[LintMessage]
-    val visitedFilterExpression = mutable.Set.empty[AnchorPosition]
-
-    var currentlyDisabledRules = Set.empty[String]
-    def trackUnusedRules(
-        rules: String,
-        toogle: Toogle,
-        anchorPosition: AnchorPosition): Unit = {
-      toogle match {
-        case Toogle.Disable =>
-          currentlyDisabledRules = currentlyDisabledRules ++ splitRules(rules)
-
-        case Toogle.Enable =>
-          val enabledRules = splitRules(rules)
-          val enabledNotDisabled = enabledRules -- currentlyDisabledRules
-
-          enabledNotDisabled.foreach(
-            rule => unusedEnable += unusedEnableWarning.at(anchorPosition.value)
-          )
+      tree.foreach {
+        case t @ Defn.Class(mods, _, _, _, _) if hasSuppressWarnings(mods) =>
```

Comment:
I've been meaning for a long time to add `mods: List[Mod]` on `Member` to avoid boilerplate like this, even more motivated now!

Reply:
Actually, could we add a `Mods` extractor in `TreeExtractors`? https://github.com/scalacenter/scalafix/blob/67a1f287618ad11144e85af3d79a3c240e0268de/scalafix-core/shared/src/main/scala/scalafix/util/TreeOps.scala#L11

```scala
object Mods {
  def unapply(tree: Tree): Option[List[Mod]] = ...
}
```

Seems like `Decl.{Val,Var,Def,Type}` are unhandled, it should be possible to annotate those as well.

Reply:
>Actually, could we add a Mods extractor in TreeExtractors?

Nice! Done

>Seems like Decl.{Val,Var,Def,Type} are unhandled, it should be possible to annotate those as well.

For some reason I missed these ones. Thanks for catching this